File: 15. 3Sum/3Sum.cpp
/*
 *@author:leacoder
 *@des:   一层枚举 左右两指针形式  三数之和 
 */
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res; //返回值
        if(nums.size()<3){ //输入检查
            return res;
        }
        sort(nums.begin(), nums.end()); //排序
        
        for (int i = 0; i < nums.size()-2; i++) //一层循环
        {
            if(i>0 && nums[i] == nums[i-1])
                continue;
            int target = -nums[i]; //其他两数和
            int l = i + 1, r = nums.size() - 1; //左右指针形式  l 左 r右
            while (l < r)
            {
                if (nums[l] + nums[r] < target) //小于目标 l 增加
                    ++l;
                else if (nums[l] + nums[r] > target) //大于目标 r 减少
                    --r;
                else{ //刚好
                    res.push_back(vector<int>{nums[i], nums[l], nums[r]});
                    while (l < r && nums[l] == nums[l+1])
                        l++;
                    while (l < r && nums[r] == nums[r-1])
                        r--;
                    l++; r--;
                }
            }
            
        }
        return res;
    }
};

File: 15. 3Sum/3Sum.py
# @author:leacoder
# @des:   借助两数之和  三数之和

class Solution(object):
    def threeSum(self, nums):
        nums.sort() #排序 方便去重
        res = []
        for i, num in enumerate(nums): #一层循环
            if i > 0 and nums[i] == nums[i-1]: # 避免重复
                continue
                
            new_nums = nums[i+1:] #剔除 一层循环后的数 
            two_sums = self.twoSum(new_nums, -num) #由于和为0  两数之和要为 -num
            for two_sum in two_sums:
                res.append([num, new_nums[two_sum[0]], new_nums[two_sum[1]]])
                         
        return res
            
        
    def twoSum(self, nums, target):
        d = {}
        res = []
        hit = False
        for i, num in enumerate(nums):
            if i > 1 and nums[i] == nums[i-1] and hit:
                continue
            if num in d:
                res.append([d[num], i])
                hit = True
            else:
                d[target - num] = i
                hit = False
        return res